File: expconslayout/src/main/java/rjsv/expconslayout/ExpandableConstraintLayout.java
package rjsv.expconslayout;

import android.animation.Animator;
import android.animation.TimeInterpolator;
import android.animation.ValueAnimator;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.constraint.ConstraintLayout;
import android.support.v4.view.animation.FastOutSlowInInterpolator;
import android.util.AttributeSet;

import rjsv.expconslayout.enumerators.ExpandableConstraintLayoutListenerStatus;
import rjsv.expconslayout.enumerators.ExpandableConstraintLayoutStatus;

import static rjsv.expconslayout.enumerators.ExpandableConstraintLayoutListenerStatus.AnimationEnd;
import static rjsv.expconslayout.enumerators.ExpandableConstraintLayoutListenerStatus.AnimationStart;
import static rjsv.expconslayout.enumerators.ExpandableConstraintLayoutListenerStatus.Closed;
import static rjsv.expconslayout.enumerators.ExpandableConstraintLayoutListenerStatus.Opened;
import static rjsv.expconslayout.enumerators.ExpandableConstraintLayoutListenerStatus.PreClose;
import static rjsv.expconslayout.enumerators.ExpandableConstraintLayoutListenerStatus.PreOpen;
import static rjsv.expconslayout.enumerators.ExpandableConstraintLayoutStatus.COLLAPSING;
import static rjsv.expconslayout.enumerators.ExpandableConstraintLayoutStatus.EXPANDING;
import static rjsv.expconslayout.enumerators.ExpandableConstraintLayoutStatus.IDLE;

/**
 * Description
 * The type Expandable linear layout.
 *
 * @author <a href="mailto:rvfts@hotmail.com"/>
 * @version $Revision : 1 $
 */

@SuppressWarnings("ResourceType")
public class ExpandableConstraintLayout extends ConstraintLayout {

    private int animationDuration;
    private boolean isVertical;
    private float expansion;
    private float displacement;
    private TimeInterpolator interpolator;
    private ExpandableConstraintLayoutListener animationListener;
    private ValueAnimator valueAnimator;
    private ExpandableConstraintLayoutStatus currentStatus = IDLE;

    // Constructors
    public ExpandableConstraintLayout(Context context) {
        this(context, null, 0);
    }

    public ExpandableConstraintLayout(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ExpandableConstraintLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        // default values
        animationDuration = 200;
        interpolator = new FastOutSlowInInterpolator();
        isVertical = true;
        displacement = 1;
        expansion = 1;
    }

    // Overridden Methods
    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        int width = getMeasuredWidth();
        int height = getMeasuredHeight();
        int size = isVertical ? height : width;
        setVisibility(expansion == 0 && size == 0 ? INVISIBLE : VISIBLE);
        int expansionDelta = size - Math.round(size * expansion);
        // translate all children before measuring the parent
        if (displacement > 0) {
            float displacementDelta = expansionDelta * displacement;
            for (int i = 0; i < getChildCount(); i++) {
                if (isVertical) {
                    getChildAt(i).setTranslationY(-displacementDelta);
                } else {
                    int direction = -1;
                    getChildAt(i).setTranslationX(direction * displacementDelta);
                }
            }
        }
        if (isVertical) {
            setMeasuredDimension(width, height - expansionDelta);
        } else {
            setMeasuredDimension(width - expansionDelta, height);
        }
    }

    // General Methods
    public void toggle() {
        if (isExpanded()) {
            collapse();
        } else {
            expand();
        }
    }

    public void expand() {
        shouldExpand(true, true);
    }

    public void collapse() {
        shouldExpand(false, true);
    }

    private void shouldExpand(boolean shouldExpand, boolean shouldAnimate) {
        if (shouldExpand && (currentStatus == EXPANDING || expansion == 1)) {
            return;
        }
        if (!shouldExpand && (currentStatus == COLLAPSING || expansion == 0)) {
            return;
        }
        int newExpansion = shouldExpand ? 1 : 0;
        if (shouldAnimate) {
            animateExpansion(newExpansion);
        } else {
            setExpansion(newExpansion);
        }
    }

    private void setExpansion(float newExpansion) {
        // Nothing to do here
        if (this.expansion == newExpansion) {
            return;
        }
        this.expansion = newExpansion;
        setVisibility(expansion == 0 ? INVISIBLE : VISIBLE);
        requestLayout();
    }

    private void animateExpansion(final int newExpansion) {
        if (valueAnimator != null) {
            valueAnimator.cancel();
            valueAnimator = null;
        }
        valueAnimator = ValueAnimator.ofFloat(expansion, newExpansion);
        valueAnimator.setInterpolator(interpolator);
        valueAnimator.setDuration(animationDuration);
        valueAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator valueAnimator) {
                setExpansion((float) valueAnimator.getAnimatedValue());
            }
        });
        valueAnimator.addListener(new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animation) {
                reportListenerStatus(AnimationStart);
                if (newExpansion > 0) {
                    reportListenerStatus(PreOpen);
                    currentStatus = EXPANDING;
                } else {
                    reportListenerStatus(PreClose);
                    currentStatus = COLLAPSING;
                }
            }

            @Override
            public void onAnimationEnd(Animator animation) {
                reportListenerStatus(AnimationEnd);
                if (currentStatus == EXPANDING) {
                    reportListenerStatus(Opened);
                } else {
                    reportListenerStatus(Closed);
                }
                currentStatus = IDLE;
            }

            @Override
            public void onAnimationCancel(Animator animation) {
                currentStatus = IDLE;
            }

            @Override
            public void onAnimationRepeat(Animator animation) {
            }
        });
        valueAnimator.start();
    }

    public boolean isExpanded() {
        return expansion == 1;
    }

    public void setInterpolator(TimeInterpolator interpolator) {
        this.interpolator = interpolator;
    }

    public TimeInterpolator getInterpolator() {
        return interpolator;
    }

    public int getAnimationDuration() {
        return animationDuration;
    }

    public void setAnimationDuration(int animationDuration) {
        this.animationDuration = animationDuration;
    }

    // Listener Related
    public void setAnimationListener(@NonNull ExpandableConstraintLayoutListener listener) {
        this.animationListener = listener;
    }

    public void reportListenerStatus(ExpandableConstraintLayoutListenerStatus status) {
        if (status != null && animationListener != null) {
            switch (status) {
                case PreOpen:
                    animationListener.onPreOpen();
                    break;
                case PreClose:
                    animationListener.onPreClose();
                    break;
                case Opened:
                    animationListener.onOpened();
                    break;
                case Closed:
                    animationListener.onClosed();
                    break;
                case AnimationStart:
                    animationListener.onAnimationStart(currentStatus);
                    break;
                case AnimationEnd:
                    animationListener.onAnimationEnd(currentStatus);
                    break;
                default:
                    break;
            }
        }
    }

}

File: expandableConstraintLayout/src/main/java/rjsv/expconslayout/ExpandableConstraintLayout.kt
package rjsv.expconslayout

import android.animation.Animator
import android.animation.TimeInterpolator
import android.animation.ValueAnimator
import android.content.Context
import android.support.constraint.ConstraintLayout
import android.support.v4.view.animation.FastOutSlowInInterpolator
import android.util.AttributeSet
import android.view.View
import rjsv.expconslayout.enumerators.ExpandableConstraintLayoutListenerStatus
import rjsv.expconslayout.enumerators.ExpandableConstraintLayoutListenerStatus.*
import rjsv.expconslayout.enumerators.ExpandableConstraintLayoutStatus.*

/**
 * Description
 * The type Expandable linear layout.
 *
 * @author [](mailto:rvfts@hotmail.com)
 * @version $Revision : 1 $
 */

class ExpandableConstraintLayout constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : ConstraintLayout(context, attrs, defStyleAttr) {

    var animationDuration: Int = 0
    private val isVertical: Boolean
    private var expansion: Float = 0.toFloat()
    private val displacement: Float
    var interpolator: TimeInterpolator? = null
    private var animationListener: ExpandableConstraintLayoutListener? = null
    private var valueAnimator: ValueAnimator? = null
    private var currentStatus = IDLE

    val isExpanded: Boolean
        get() = expansion == 1f

    init {
        // default values
        animationDuration = 200
        interpolator = FastOutSlowInInterpolator()
        isVertical = true
        displacement = 1f
        expansion = 1f
    }

    // Overridden Methods
    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val width = measuredWidth
        val height = measuredHeight
        val size = if (isVertical) height else width
        visibility = if (expansion == 0f && size == 0) View.INVISIBLE else View.VISIBLE
        val expansionDelta = size - Math.round(size * expansion)
        // translate all children before measuring the parent
        if (displacement > 0) {
            val displacementDelta = expansionDelta * displacement
            for (i in 0 until childCount) {
                if (isVertical) {
                    getChildAt(i).translationY = -displacementDelta
                } else {
                    val direction = -1
                    getChildAt(i).translationX = direction * displacementDelta
                }
            }
        }
        if (isVertical) {
            setMeasuredDimension(width, height - expansionDelta)
        } else {
            setMeasuredDimension(width - expansionDelta, height)
        }
    }

    // General Methods
    fun toggle() {
        if (isExpanded) {
            collapse()
        } else {
            expand()
        }
    }

    fun expand() {
        shouldExpand(true, true)
    }

    fun collapse() {
        shouldExpand(false, true)
    }

    private fun shouldExpand(shouldExpand: Boolean, shouldAnimate: Boolean) {
        if (shouldExpand && (currentStatus == EXPANDING || expansion == 1f)) {
            return
        }
        if (!shouldExpand && (currentStatus == COLLAPSING || expansion == 0f)) {
            return
        }
        val newExpansion = if (shouldExpand) 1F else 0F
        if (shouldAnimate) {
            animateExpansion(newExpansion)
        } else {
            setExpansion(newExpansion)
        }
    }

    private fun setExpansion(newExpansion: Float) {
        // Nothing to do here
        if (this.expansion == newExpansion) {
            return
        }
        this.expansion = newExpansion
        visibility = if (expansion == 0f) View.INVISIBLE else View.VISIBLE
        requestLayout()
    }

    private fun animateExpansion(newExpansion: Float) {
        if (valueAnimator != null) {
            valueAnimator?.cancel()
            valueAnimator = null
        }
        valueAnimator = ValueAnimator.ofFloat(expansion, newExpansion)
        valueAnimator?.interpolator = interpolator
        valueAnimator?.duration = animationDuration.toLong()
        valueAnimator?.addUpdateListener { valueAnimator -> setExpansion(valueAnimator.animatedValue as Float) }
        valueAnimator?.addListener(object : Animator.AnimatorListener {
            override fun onAnimationStart(animation: Animator) {
                reportListenerStatus(AnimationStart)
                currentStatus = if (newExpansion <= 0) {
                    reportListenerStatus(PreClose)
                    COLLAPSING
                } else {
                    reportListenerStatus(PreOpen)
                    EXPANDING
                }
            }

            override fun onAnimationEnd(animation: Animator) {
                reportListenerStatus(AnimationEnd)
                if (currentStatus == EXPANDING) {
                    reportListenerStatus(Opened)
                } else {
                    reportListenerStatus(Closed)
                }
                currentStatus = IDLE
            }

            override fun onAnimationCancel(animation: Animator) {
                currentStatus = IDLE
            }

            override fun onAnimationRepeat(animation: Animator) {}
        })
        valueAnimator?.start()
    }

    // Listener Related
    fun setAnimationListener(listener: ExpandableConstraintLayoutListener) {
        this.animationListener = listener
    }

    fun reportListenerStatus(status: ExpandableConstraintLayoutListenerStatus?) {
        if (status != null && animationListener != null) {
            when (status) {
                PreOpen -> animationListener!!.onPreOpen()
                PreClose -> animationListener!!.onPreClose()
                Opened -> animationListener!!.onOpened()
                Closed -> animationListener!!.onClosed()
                AnimationStart -> animationListener!!.onAnimationStart(currentStatus)
                AnimationEnd -> animationListener!!.onAnimationEnd(currentStatus)
                else -> {
                }
            }
        }
    }

}// Constructors
